File: plugin/vscode/src/main/settings/Options.ts
import * as vscode from "vscode";
import { ClassCheckMode, classCheckFlag } from "./ClassCheckMode";
import { toCliPath } from "./ProjectPaths";

type ArgSink = (value: string) => void;

/** Map workspace settings to docsig CLI arguments. */
export class Options {
  constructor(private readonly folder?: vscode.WorkspaceFolder) {}

  addArgs(add: ArgSink): void {
    const config = vscode.workspace.getConfiguration(
      "docsig",
      this.folder?.uri,
    );

    const classMode = config.get<ClassCheckMode>(
      "classCheckMode",
      ClassCheckMode.None,
    );
    const classFlag = classCheckFlag(classMode);
    if (classFlag) {
      add(classFlag);
    }

    this.addBool(config, "checkDunders", "--check-dunders", add);
    this.addBool(config, "checkNested", "--check-nested", add);
    this.addBool(config, "checkOverridden", "--check-overridden", add);
    this.addBool(
      config,
      "checkPropertyReturns",
      "--check-property-returns",
      add,
    );
    this.addBool(config, "checkProtected", "--check-protected", add);
    this.addBool(
      config,
      "checkProtectedClassMethods",
      "--check-protected-class-methods",
      add,
    );
    this.addBool(config, "ignoreArgs", "--ignore-args", add);
    this.addBool(config, "ignoreKwargs", "--ignore-kwargs", add);
    this.addBool(config, "ignoreNoParams", "--ignore-no-params", add);
    this.addBool(config, "includeIgnored", "--include-ignored", add);

    const exclude = config.get<string>("exclude", "")?.trim();
    if (exclude) {
      add("--exclude");
      add(exclude);
    }

    const excludes = config.get<string[]>("excludes", []).filter(Boolean);
    if (excludes.length > 0) {
      add("--excludes");
      excludes.forEach((entry) =>
        add(toCliPath(this.folder?.uri.fsPath, entry)),
      );
    }

    const disable = config.get<string[]>("disable", []).filter(Boolean);
    if (disable.length > 0) {
      add("--disable");
      add(disable.join(","));
    }

    const target = config.get<string[]>("target", []).filter(Boolean);
    if (target.length > 0) {
      add("--target");
      add(target.join(","));
    }
  }

  private addBool(
    config: vscode.WorkspaceConfiguration,
    key: string,
    flag: string,
    add: ArgSink,
  ): void {
    if (config.get<boolean>(key, false)) {
      add(flag);
    }
  }
}

File: editors/neovim/lua/docsig/options.lua
local class_check_mode = require("docsig.class_check_mode")
local project_paths = require("docsig.project_paths")

local M = {}

local function add_bool(cfg, key, flag, args)
  if cfg[key] then args[#args + 1] = flag end
end

function M.build_args(cfg, root)
  local args = {}

  local class_flag = class_check_mode.flag(cfg.class_check_mode)
  if class_flag then args[#args + 1] = class_flag end

  add_bool(cfg, "check_dunders", "--check-dunders", args)
  add_bool(cfg, "check_nested", "--check-nested", args)
  add_bool(cfg, "check_overridden", "--check-overridden", args)
  add_bool(cfg, "check_property_returns", "--check-property-returns", args)
  add_bool(cfg, "check_protected", "--check-protected", args)
  add_bool(
    cfg,
    "check_protected_class_methods",
    "--check-protected-class-methods",
    args
  )
  add_bool(cfg, "ignore_args", "--ignore-args", args)
  add_bool(cfg, "ignore_kwargs", "--ignore-kwargs", args)
  add_bool(cfg, "ignore_no_params", "--ignore-no-params", args)
  add_bool(cfg, "include_ignored", "--include-ignored", args)

  local exclude = vim.trim(cfg.exclude or "")
  if exclude ~= "" then
    args[#args + 1] = "--exclude"
    args[#args + 1] = exclude
  end

  local excludes = cfg.excludes or {}
  if #excludes > 0 then
    args[#args + 1] = "--excludes"
    for _, entry in ipairs(excludes) do
      if entry and entry ~= "" then
        args[#args + 1] = project_paths.to_cli_path(root, entry)
      end
    end
  end

  local disable = cfg.disable or {}
  if #disable > 0 then
    args[#args + 1] = "--disable"
    args[#args + 1] = table.concat(disable, ",")
  end

  local target = cfg.target or {}
  if #target > 0 then
    args[#args + 1] = "--target"
    args[#args + 1] = table.concat(target, ",")
  end

  return args
end

return M
